### PR TITLE
[release/6.0] Fixed the yp_spin_count_unit to be a factor of the original_spin_count_unit rather than a continually increasing value 

### PR DESCRIPTION
**Pull request by mrsharm**
Backport of https://github.com/dotnet/runtime/pull/68879 to release/6.0. 

## Customer Impact

Reported by Cosmos; the associated bug can be found here: https://github.com/dotnet/runtime/issues/68839  

## Details

This PR is to address https://github.com/dotnet/runtime/issues/68839 where it's clear that ``SetYieldProcessorScalingFactor`` is called multiple times thereby increasing the value of ``yp_spin_count_unit`` and as a result the spin counts are monotonically increasing each time ``SetYieldProcessorScalingFactor`` is called; this function was previously assumed to be called only once but that changed with the following PR: https://github.com/dotnet/runtime/pull/55295.

The fix is to store the original spin count unit and use that value to set the ``yp_spin_count_unit`` rather than adjusting the value based on the last ``yp_spin_count_unit``. 

## Testing

Tested on a simple console app that has Server GC enabled and made sure the spin lock count is constrained. The repro code as a simple console app is:

```csharp
var ALLOC_SIZE = 10_000;

while (true)
{
    Thread.SpinWait(10000);
    byte[] buffer = new byte[ALLOC_SIZE];
    buffer = new byte[ALLOC_SIZE];
    buffer = new byte[ALLOC_SIZE];
}
```

``Thread.SpinWait`` eventually calls ``SetYieldProcessorScalingFactor``; the result of the testing indicated that the ``yp_spin_count`` wasn't monotonically increasing like before. 

## Risk

This PR mitigates a high-risk issue and reduces the risk of the spin count growing by constraining max count.

### CODE REVIEW

File: src/coreclr/gc/gc.cpp [hunk 1]
```diff
@@ -43038,10 +43041,10 @@ void GCHeap::SetYieldProcessorScalingFactor (float scalingFactor)
 {
     assert (yp_spin_count_unit != 0);
     int saved_yp_spin_count_unit = yp_spin_count_unit;
-    yp_spin_count_unit = (int)((float)yp_spin_count_unit * scalingFactor / (float)9);
+    yp_spin_count_unit = (int)((float)original_spin_count_unit * scalingFactor / (float)9);
 
-    // It's very suspicious if it becomes 0
-    if (yp_spin_count_unit == 0)
+    // It's very suspicious if it becomes 0 and also, we don't want to spin too much.
+    if ((yp_spin_count_unit == 0) || (yp_spin_count_unit > 32768))
```
Comment (mangod9) [inline src/coreclr/gc/gc.cpp:43047]: why max of 32k here?

**Review by jeffschwMSFT — APPROVED** [review]
Approved.  Please fill in the servicing template, and get a handful of code reviews.  We can then take for consideration in 6.0.x

File: src/coreclr/gc/gc.cpp (hunk 1 above)
Comment (mrsharm) [reply src/coreclr/gc/gc.cpp:43047]: As Per, @Maoni0's comment on https://github.com/dotnet/runtime/pull/68879#issuecomment-1118093244: 

> I would put a cap on the check too 'cause we don't want to spin too much -
> 
> ` if ((yp_spin_count_unit == 0) || (yp_spin_count_unit > 32768))`

This is to hedge our bets against a large spin count.

Comment (Maoni0) [reply src/coreclr/gc/gc.cpp:43047]: we do 32*number_of_heaps by default. so this is saying if you had 1024 heaps, which would be very many heaps.

Comment (mangod9) [reply src/coreclr/gc/gc.cpp:43047]: hmm.. ok

**Review by mangod9 — APPROVED** [review]

**Review by kouvel — APPROVED** [review]

**Pull request merged**